src/HIE/Bios/Cradle.hs

```diff
@@ -353,39 +362,68 @@ cabalWorkDir = findFileUpwards isCabal
 -- Stack Cradle
 -- Works for by invoking `stack repl` with a wrapper script
 
-stackCradle :: FilePath -> Cradle
-stackCradle wdir =
+stackCradle :: FilePath -> Maybe String -> Cradle
+stackCradle wdir mc =
   Cradle
     { cradleRootDir    = wdir
     , cradleOptsProg   = CradleAction
         { actionName = "stack"
-        , runCradle = stackAction wdir
+        , runCradle = stackAction wdir mc
         }
     }
 
-stackCradleDependencies :: FilePath -> IO [FilePath]
+stackCradleDependencies :: FilePath-> IO [FilePath]
 stackCradleDependencies wdir = do
   cabalFiles <- findCabalFiles wdir
   return $ cabalFiles ++ ["package.yaml", "stack.yaml"]
 
-stackAction :: FilePath -> LoggingFunction -> FilePath -> IO (CradleLoadResult ComponentOptions)
-stackAction work_dir l fp = do
+stackAction :: FilePath -> Maybe String -> LoggingFunction -> FilePath -> IO (CradleLoadResult ComponentOptions)
+stackAction work_dir mc l _fp = do
+  ghcPath <- readCreateProcess
+    (readProcessInDirectory work_dir "stack" ["path", "--compiler-exe"])
+    ""
   -- Same wrapper works as with cabal
-  wrapper_fp <- getCabalWrapperTool
+  wrapper_fp <- getCabalWrapperTool (return ghcPath)
+
   (ex1, _stdo, stde, args) <-
-      readProcessWithOutputFile l work_dir "stack" ["repl", "--no-nix-pure", "--no-load", "--with-ghc", wrapper_fp, fp ]
+    readProcessWithOutputFile
+            l (Just ghcPath) work_dir
+            "stack" $ ["repl", "--no-nix-pure", "--with-ghc", wrapper_fp] ++ maybeToList mc
   (ex2, pkg_args, stdr, _) <-
-    readProcessWithOutputFile l work_dir "stack" ["path", "--ghc-package-path"]
+    readProcessWithOutputFile l Nothing work_dir "stack" ["path", "--ghc-package-path"]
   let split_pkgs = concatMap splitSearchPath pkg_args
       pkg_ghc_args = concatMap (\p -> ["-package-db", p] ) split_pkgs
   deps <- stackCradleDependencies work_dir
-  return $ case processCabalWrapperArgs args of
-      Nothing -> CradleFail (CradleError ex1 $
+  case processCabalWrapperArgs args of
+      Nothing -> return $ CradleFail (CradleError ex1 $
                   ("Failed to parse result of calling cabal":
                     stde)
                    ++ args)
 
-      Just ghc_args -> makeCradleResult (combineExitCodes [ex1, ex2], stde ++ stdr, ghc_args ++ pkg_ghc_args) deps
+      Just ghc_args -> do
+        -- parse the .ghci file that contains all targets
+        -- a package declares.
+        let Just fp = findGhciScript ghc_args
+        targets <- getTargetsFromGhciScript fp
+        return $ makeCradleResult (combineExitCodes [ex1, ex2], stde ++ stdr, ghc_args ++ pkg_ghc_args ++ targets) deps
+
+findGhciScript :: [String] -> Maybe FilePath
+findGhciScript args
+    | Just option <- find (ghciScript `isPrefixOf`) args
+    , Just flag <- stripPrefix ghciScript option
+    = Just flag
+    | otherwise = Nothing
+    where
+      ghciScript = "-ghci-script="
```

Comment:
This is a bit dodgy. I though this logic would be in the session initialization function as then you can look at the `DynFlags` field to get the ghci script file just before setting the targets.

Reply:
I felt like here it would also be nice because we can easily debug it. It would follow the principle that the output of `hie-bios flags ...` can directly be used to compile a file.
If we are looking at the DynFlags in initSession for example, then we dont know the cradle type anymore and parse the ghci-script for every cradle. If this is desired behaviour, then it makes sense.